### ISSUE TITLE
Undefined offset: 0 on ModelWatcher.php 

### ISSUE THREAD

**Issue by ismailcaakir** [laravel/telescope#187]
Same error, an error when updating my database

http://prntscr.com/laqbwg

**Comment by FrancoisDucobu** [comment]
We got the same issue...

If I make a dump($data) before line 35 on ModelWatcher.php, I see that $data does not contain an array but the a model.

To avoid this error, on DatabaseSeeder.php, just before the seed, we call this command

config('telescope.watchers.Watchers\ModelWatcher',false);

And we call this at the end of the seed.

config('telescope.watchers.Watchers\ModelWatcher',true);

But I would like to say that we like telescope a lot ! Keep it going ! :D

**Comment by laypassmail** [comment]
I got the same issue when call create function via morphMany relationship

**Comment by jacobzlogar** [comment]
Still working on steps to reproduce, but yeah as @laypassmail stated, the array has a key of "model", most of the places i was able to replicate this were through dispatched jobs.. will update when i have better steps to reproduce
<img width="307" alt="screen shot 2018-10-30 at 10 23 11 am" src="https://user-images.githubusercontent.com/10762368/47724967-17297780-dc2e-11e8-9d02-4fda2eaafa8c.png">

**Comment by joselara** [comment]
I'm having the same issue and need to remove from my local machine.

**Comment by jacobzlogar** [comment]
This is caused by using this package https://github.com/GeneaLabs/laravel-model-caching, which has a dependency of https://github.com/fico7489/laravel-pivot, this package adds events for pivot relations, this line here https://github.com/fico7489/laravel-pivot/blob/master/src/Traits/ExtendFireModelEventTrait.php#L34 is causing the array keys to not match in the `recordAction` method of `ModelWatcher`

**Comment by laypassmail** [comment]
thank @jacobzlogar, but it still get an error, pls check again

**Comment by jacobzlogar** [comment]
@laypassmail i don't think my changes made it into the 0.1.6 release

**Comment by laypassmail** [comment]
@jacobzlogar, but i see in v0.1.6 https://github.com/laravel/telescope/compare/v0.1.4...v0.1.6#diff-b61a0d5e3e3abd5547f3b99ada5d9e9bR65

**Comment by jacobzlogar** [comment]
@laypassmail https://github.com/laravel/telescope/pull/282/files#diff-b61a0d5e3e3abd5547f3b99ada5d9e9bR35 The changes i proposed in this PR (using `current($data);` to find the first index of the event payload) would've solved this issue, but i'm not sure if its the best solution possibly @themsaid has input?

**Comment by laypassmail** [comment]
@joselara , I saw that, i think larave-pilot will replease v 4.*, It would solve this issue https://github.com/fico7489/laravel-pivot/pull/50#issuecomment-435970639

**Comment by OwenMelbz** [comment]
Just to confirm - we can also produce this, we're using `genealabs/laravel-model-caching` - however it does seem maybe a bit over zealous to assume `$item[0]` will be correct? would something less delicate like `current($item)` be suitable replacement?

**Comment by themsaid** [comment]
How can I replicate it?

**Comment by OwenMelbz** [comment]
Hey @themsaid 

It *might* just be incompatibility with caching packages that used named array keys.

The way we can replicate it is by having `genealabs/laravel-model-caching` installed - then making various changes to the model and saving it.

It happens for us in a pattern of

- fails
- fails
- works
- fails
- fails
- works

--- I know this is a weird way round to solve it, but instead of trying to replicate it, would it be worth just refactoring the method so it works without using `[0]` ? Then as long as the tests still pass, we know it should be okay?

Additionally you could for the sake of testing - before getting to that method, change the array key name manually to test it works?

**Comment by themsaid** [comment]
All Eloquent model events has the model at index zero of the payload, I don't see how this may not be the case.

**Comment by OwenMelbz** [comment]
I wouldn't say they were "breaking the core" - as they're fulfilling a requirement that Laravel is "officially" lacking, e.g. Model Caching and A full suite of eloquent events.

Let's say you guys decided to add all the required eloquent events into the core - then the package wouldn't be needed, same would go for model caching.

Anything Laravel doesn't provide - people will have to provide community solutions too, and those are popular solutions - it would seem most noble to help support that community by making sure popular packages do not break because they're trying to help others in an unofficial way.

I admit I don't know *why* they're doing them that way, maybe they should be updated to be compatible - but at the same time, if the Laravel core could make sure the code in this instance is a little more forgiving - it would help support thousands of code-bases that have these installed, at no detriment to the code base?

For example we've probably spent more time discussing this than it would have taken to provide a solution that solves it for everybody!

I guess you could think of it like... Instead of treating millions for a sickness - eradicate the sickness at the core so you don't need to treat the millions :D

**Comment by themsaid** [comment]
@OwenMelbz I'm not against the packages, I just don't think doing it this way is correct.

When I'm building a package the hardest and most important part is to keep the core un-touched, a user can simply remove my package and his app should work fine after removing any references to the package.

Anyway, feel free to open a PR.

**Comment by drbyte** [comment]
@themsaid it's in a Controller's `store` method:

```php
$rules = [ array of validation rules];
$this->validate($request, $rules);

$data            = [
    'first'      => $request->input('m_first'),
    'last'       => $request->input('m_last'),
    'cellphone'  => $request->input('m_cellphone'),
    'email'      => $request->input('m_email'),
    'username'   => $request->input('m_username'),
    //
    ];
$user = new \App\User($data);

$user->created_by = $request->user()->name;

$user->active                     = false;
$user->allow_address_share        = false;
$user->receive_email_news_general = false;

$user->save();
```

**Comment by themsaid** [comment]
I don't see the `UserAdded` event fired in the code you shared @drbyte

**Comment by mikebronner** [comment]
I'm just adding myself here to keep updated on this discussion, in case Laravel-model-caching needs to be updated. From what I gather so far its an issue that occurs in laravel-pivot, but want to keep an eye on it if there is anything I can do in my package.

**Comment by themsaid** [comment]
@mikebronner Model events must have the model instance at index zero of the event payload, just like how the core works.

**Comment by mikebronner** [comment]
@themsaid Thanks ... yes, my package is not firing events, so it is likely in laravel-pivot, as that package does fire events, and I am using it as a dependency to track when pivots are updated, so I can listen for that.

**Comment by themsaid** [comment]
@drbyte no, my comment is only for model events. Yours is a custom event, I've already added a fix to Telescope for your case and should be released tomorrow.

**Comment by drbyte** [comment]
Thanks @themsaid

**Comment by paras-malhotra** [comment]
#465 will fix the model event issue. See https://github.com/laravel/telescope/issues/351#issuecomment-443523197

**Comment by SirLamer** [comment]
Here is an alternative implementation of the custom watcher which does not replace the existing Telescope code so that it will be more resilient through Telescope code base changes.

```php
<?php

namespace App\Watchers;

use \Laravel\Telescope\Watchers\ModelWatcher as PackageWatcher;

class ModelWatcher extends PackageWatcher {

    public function recordAction($event, $data)
    {
        //Catch and convert model events from some non-framework model event implementations such as
        //genealabs/laravel-model-caching and fico7489/laravel-pivot
        //Tested with fico7489/laravel-pivot v3.0.0 on 2019-02-02
        $newData = $data;
        $newData[0] = $newData[0] ?? $newData['model'];

        parent::recordAction($event, $newData);
    }
}
```

**Comment by braulioholtz** [comment]
> Here is an alternative implementation of the custom watcher which does not replace the existing Telescope code so that it will be more resilient through Telescope code base changes.
> 
> ```
> use \Laravel\Telescope\Watchers\ModelWatcher as PackageWatcher;
> 
> class ModelWatcher extends PackageWatcher {
> 
>     public function recordAction($event, $data)
>     {
>         //Catch and convert model events from some non-framework model event implementations such as
>         //genealabs/laravel-model-caching and fico7489/laravel-pivot
>         //Tested with fico7489/laravel-pivot v3.0.0 on 2019-02-02
>         $modifiedData = $data;
>         $modifiedData[0] = $modifiedData[0] ?? $modifiedData['model'];
> 
>         parent::recordAction($event, $modifiedData);
>     }
> }
> ```

How did you create this file and how do you make the call?

**Comment by Ben52** [comment]
@braulioholtz You can create that class anywhere you like, I created it in `app`, and in `config/telescope.php`, in the watchers array, replace `Watchers\ModelWatcher::class => env('TELESCOPE_MODEL_WATCHER', true)` with your custom ModelWatcher class.

**Comment by SirLamer** [comment]
@braulioholtz I apologize, I posted this comment to the wrong GitHub issue! It was meant to be posted here: https://github.com/fico7489/laravel-pivot/issues/55

I am going to post the same there as well.
